// eas-cli/build/commands/device/create.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const AppStoreApi_1 = tslib_1.__importDefault(require("../../credentials/ios/appstore/AppStoreApi"));
const context_1 = require("../../devices/context");
const manager_1 = tslib_1.__importDefault(require("../../devices/manager"));
class DeviceCreate extends EasCommand_1.default {
    async runAsync() {
        // this command is interactive by design
        const { loggedIn: { actor, graphqlClient }, privateProjectConfig, } = await this.getContextAsync(_a, {
            nonInteractive: false,
        });
        const ctx = await (0, context_1.createContextAsync)({
            appStore: new AppStoreApi_1.default(),
            user: actor,
            graphqlClient,
            projectId: privateProjectConfig?.projectId,
        });
        const manager = new manager_1.default(ctx);
        await manager.createAsync();
    }
}
_a = DeviceCreate;
DeviceCreate.description = 'register new Apple Devices to use for internal distribution';
DeviceCreate.contextDefinition = {
    ..._a.ContextOptions.LoggedIn,
    ..._a.ContextOptions.OptionalProjectConfig,
};
exports.default = DeviceCreate;
