// eas-cli/build/commandUtils/context/MaybeLoggedInContextField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const ContextField_1 = tslib_1.__importDefault(require("./ContextField"));
const createGraphqlClient_1 = require("./contextUtils/createGraphqlClient");
const FeatureGateEnvOverrides_1 = tslib_1.__importDefault(require("../gating/FeatureGateEnvOverrides"));
const FeatureGating_1 = tslib_1.__importDefault(require("../gating/FeatureGating"));
class MaybeLoggedInContextField extends ContextField_1.default {
    async getValueAsync({ sessionManager }) {
        const authenticationInfo = {
            accessToken: sessionManager.getAccessToken(),
            sessionSecret: sessionManager.getSessionSecret(),
        };
        const graphqlClient = (0, createGraphqlClient_1.createGraphqlClient)(authenticationInfo);
        const actor = await sessionManager.getUserAsync();
        const featureGateServerValues = actor?.featureGates ?? {};
        return {
            actor: actor ?? null,
            featureGating: new FeatureGating_1.default(featureGateServerValues, new FeatureGateEnvOverrides_1.default()),
            graphqlClient,
        };
    }
}
exports.default = MaybeLoggedInContextField;
