// eas-cli/build/commands/build/version/set.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Updates_1 = require("@expo/config-plugins/build/utils/Updates");
const eas_build_job_1 = require("@expo/eas-build-job");
const eas_json_1 = require("@expo/eas-json");
const core_1 = require("@oclif/core");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const EasCommand_1 = tslib_1.__importDefault(require("../../../commandUtils/EasCommand"));
const AppVersionMutation_1 = require("../../../graphql/mutations/AppVersionMutation");
const AppVersionQuery_1 = require("../../../graphql/queries/AppVersionQuery");
const AppPlatform_1 = require("../../../graphql/types/AppPlatform");
const log_1 = tslib_1.__importDefault(require("../../../log"));
const platform_1 = require("../../../platform");
const versions_1 = require("../../../project/android/versions");
const applicationIdentifier_1 = require("../../../project/applicationIdentifier");
const versions_2 = require("../../../project/ios/versions");
const projectUtils_1 = require("../../../project/projectUtils");
const remoteVersionSource_1 = require("../../../project/remoteVersionSource");
const prompts_1 = require("../../../prompts");
class BuildVersionSetView extends EasCommand_1.default {
    async runAsync() {
        const { flags } = await this.parse(_a);
        const { loggedIn: { graphqlClient }, getDynamicPrivateProjectConfigAsync, projectDir, vcsClient, } = await this.getContextAsync(_a, {
            nonInteractive: false,
        });
        const platform = await (0, platform_1.selectPlatformAsync)(flags.platform);
        const easJsonAccessor = eas_json_1.EasJsonAccessor.fromProjectPath(projectDir);
        await (0, remoteVersionSource_1.ensureVersionSourceIsRemoteAsync)(easJsonAccessor, { nonInteractive: false });
        const profile = await eas_json_1.EasJsonUtils.getBuildProfileAsync(easJsonAccessor, platform, flags.profile ?? undefined);
        const { exp, projectId } = await getDynamicPrivateProjectConfigAsync({ env: profile.env });
        const displayName = await (0, projectUtils_1.getDisplayNameForProjectIdAsync)(graphqlClient, projectId);
        (0, remoteVersionSource_1.validateAppConfigForRemoteVersionSource)(exp, platform);
        const applicationIdentifier = await (0, applicationIdentifier_1.getApplicationIdentifierAsync)({
            graphqlClient,
            projectDir,
            projectId,
            exp,
            buildProfile: profile,
            platform,
            vcsClient,
            nonInteractive: false,
        });
        const remoteVersions = await AppVersionQuery_1.AppVersionQuery.latestVersionAsync(graphqlClient, projectId, (0, AppPlatform_1.toAppPlatform)(platform), applicationIdentifier);
        const currentStateMessage = remoteVersions?.buildVersion
            ? `Project ${chalk_1.default.bold(displayName)} with ${getApplicationIdentifierName(platform)} "${applicationIdentifier}" is configured with ${(0, remoteVersionSource_1.getBuildVersionName)(platform)} ${remoteVersions.buildVersion}.`
            : `Project ${chalk_1.default.bold(displayName)} with ${getApplicationIdentifierName(platform)} "${applicationIdentifier}" does not have any ${(0, remoteVersionSource_1.getBuildVersionName)(platform)} configured.`;
        const versionPromptMessage = remoteVersions?.buildVersion
            ? `What version would you like to set?`
            : `What version would you like to initialize it with?`;
        log_1.default.log(currentStateMessage);
        const { version } = await (0, prompts_1.promptAsync)({
            type: platform === eas_build_job_1.Platform.ANDROID ? 'number' : 'text',
            name: 'version',
            message: versionPromptMessage,
            validate: platform === eas_build_job_1.Platform.ANDROID
                ? value => (0, versions_1.isValidVersionCode)(value) || `Invalid value: ${versions_1.VERSION_CODE_REQUIREMENTS}.`
                : value => (0, versions_2.isValidBuildNumber)(value) || `Invalid value: ${versions_2.BUILD_NUMBER_REQUIREMENTS}.`,
        });
        await AppVersionMutation_1.AppVersionMutation.createAppVersionAsync(graphqlClient, {
            appId: projectId,
            platform: (0, AppPlatform_1.toAppPlatform)(platform),
            applicationIdentifier,
            storeVersion: exp.version ?? '1.0.0',
            buildVersion: String(version),
            runtimeVersion: (await (0, Updates_1.getRuntimeVersionNullableAsync)(projectDir, exp, platform)) ?? undefined,
        });
    }
}
_a = BuildVersionSetView;
BuildVersionSetView.description = 'update version of an app';
BuildVersionSetView.flags = {
    platform: core_1.Flags.enum({
        char: 'p',
        options: ['android', 'ios'],
    }),
    profile: core_1.Flags.string({
        char: 'e',
        description: 'Name of the build profile from eas.json. Defaults to "production" if defined in eas.json.',
        helpValue: 'PROFILE_NAME',
    }),
};
BuildVersionSetView.contextDefinition = {
    ..._a.ContextOptions.LoggedIn,
    ..._a.ContextOptions.DynamicProjectConfig,
    ..._a.ContextOptions.ProjectDir,
    ..._a.ContextOptions.Vcs,
};
exports.default = BuildVersionSetView;
function getApplicationIdentifierName(platform) {
    if (platform === eas_build_job_1.Platform.ANDROID) {
        return 'application ID';
    }
    else {
        return 'bundle identifier';
    }
}
