// eas-cli/build/credentials/ios/actions/SetUpInternalProvisioningProfile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetUpInternalProvisioningProfile = void 0;
const tslib_1 = require("tslib");
const BuildCredentialsUtils_1 = require("./BuildCredentialsUtils");
const SetUpAdhocProvisioningProfile_1 = require("./SetUpAdhocProvisioningProfile");
const SetUpProvisioningProfile_1 = require("./SetUpProvisioningProfile");
const generated_1 = require("../../../graphql/generated");
const log_1 = tslib_1.__importStar(require("../../../log"));
const prompts_1 = require("../../../prompts");
class SetUpInternalProvisioningProfile {
    constructor(options) {
        this.options = options;
    }
    async runAsync(ctx) {
        const buildCredentials = await (0, BuildCredentialsUtils_1.getAllBuildCredentialsAsync)(ctx, this.options.app);
        const adhocBuildCredentialsExist = buildCredentials.filter(({ iosDistributionType }) => iosDistributionType === generated_1.IosDistributionType.AdHoc).length > 0;
        const enterpriseBuildCredentialsExist = buildCredentials.filter(({ iosDistributionType }) => iosDistributionType === generated_1.IosDistributionType.Enterprise).length > 0;
        if (!ctx.nonInteractive) {
            if (ctx.appStore.authCtx) {
                if (ctx.appStore.authCtx.team.inHouse) {
                    return await this.askForDistributionTypeAndSetupAsync(ctx, 'Which credentials would you like to set up?');
                }
                else {
                    return await this.setupAdhocProvisioningProfileAsync(ctx);
                }
            }
            else {
                if (adhocBuildCredentialsExist && enterpriseBuildCredentialsExist) {
                    log_1.default.log('You have set up both adhoc and universal distribution credentials.');
                    return await this.askForDistributionTypeAndSetupAsync(ctx, 'Which credentials would you like to use?');
                }
                else if (adhocBuildCredentialsExist) {
                    return await this.setupAdhocProvisioningProfileAsync(ctx);
                }
                else if (enterpriseBuildCredentialsExist) {
                    return await this.setupUniversalProvisioningProfileAsync(ctx);
                }
                else {
                    log_1.default.addNewLineIfNone();
                    log_1.default.log('You need to log in to your Apple Developer account to generate credentials for internal distribution builds, or provide credentials via credentials.json');
                    log_1.default.log((0, log_1.learnMore)('https://docs.expo.dev/app-signing/local-credentials/'));
                    const { team } = await ctx.appStore.ensureAuthenticatedAsync();
                    if (team.inHouse) {
                        return await this.askForDistributionTypeAndSetupAsync(ctx, 'Which credentials would you like to set up?');
                    }
                    else {
                        return await this.setupAdhocProvisioningProfileAsync(ctx);
                    }
                }
            }
        }
        else {
            if (adhocBuildCredentialsExist && enterpriseBuildCredentialsExist) {
                throw new Error(`You're in non-interactive mode. You have set up both adhoc and universal distribution credentials. Set the 'enterpriseProvisioning' property (to 'adhoc' or 'universal') in eas.json to choose the credentials to use.`);
            }
            else if (adhocBuildCredentialsExist) {
                return await this.setupAdhocProvisioningProfileAsync(ctx);
            }
            else if (enterpriseBuildCredentialsExist) {
                return await this.setupUniversalProvisioningProfileAsync(ctx);
            }
            else {
                throw new Error(`You're in non-interactive mode. EAS CLI couldn't find any credentials suitable for internal distribution. Run this command again in interactive mode.`);
            }
        }
    }
    async setupAdhocProvisioningProfileAsync(ctx) {
        const { app, target } = this.options;
        return await new SetUpAdhocProvisioningProfile_1.SetUpAdhocProvisioningProfile({ app, target }).runAsync(ctx);
    }
    async setupUniversalProvisioningProfileAsync(ctx) {
        return await new SetUpProvisioningProfile_1.SetUpProvisioningProfile(this.options.app, this.options.target, generated_1.IosDistributionType.Enterprise).runAsync(ctx);
    }
    async askForDistributionTypeAndSetupAsync(ctx, message) {
        const { distributionType } = await (0, prompts_1.promptAsync)({
            type: 'select',
            name: 'distributionType',
            message,
            choices: [
                { title: 'Universal Distribution', value: generated_1.IosDistributionType.Enterprise },
                { title: 'Adhoc Distribution', value: generated_1.IosDistributionType.AdHoc },
            ],
        });
        if (distributionType === generated_1.IosDistributionType.Enterprise) {
            return await this.setupUniversalProvisioningProfileAsync(ctx);
        }
        else {
            return await this.setupAdhocProvisioningProfileAsync(ctx);
        }
    }
}
exports.SetUpInternalProvisioningProfile = SetUpInternalProvisioningProfile;
