// react-native-mask-text/lib/commonjs/components/MaskedTextInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MaskedTextInputComponent = exports.MaskedTextInput = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _mask = require("../utils/mask");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const MaskedTextInputComponent = (_ref, ref) => {
  let {
    mask: pattern = '',
    type = 'custom',
    options = {},
    defaultValue,
    onChangeText,
    inputAccessoryView,
    ...rest
  } = _ref;
  const defaultValueCustom = defaultValue || '';
  const defaultValueCurrency = defaultValue || '0';
  const initialMaskedValue = type === 'currency' ? (0, _mask.mask)(defaultValueCurrency, pattern, type, options) : (0, _mask.mask)(defaultValueCustom, pattern, type, options);
  const initialUnMaskedValue = type === 'currency' ? (0, _mask.unMask)(defaultValueCurrency, type) : (0, _mask.unMask)(defaultValueCustom, type);
  const [maskedValue, setMaskedValue] = (0, _react.useState)(initialMaskedValue);
  const [unMaskedValue, setUnmaskedValue] = (0, _react.useState)(initialUnMaskedValue);

  function onChange(value) {
    const newUnMaskedValue = (0, _mask.unMask)(value, type);
    const newMaskedValue = (0, _mask.mask)(newUnMaskedValue, pattern, type, options);
    setMaskedValue(newMaskedValue);
    setUnmaskedValue(newUnMaskedValue);
  }

  (0, _react.useEffect)(() => {
    onChangeText(maskedValue, unMaskedValue);
  }, [maskedValue, unMaskedValue]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactNative.TextInput, _extends({
    onChangeText: value => onChange(value),
    ref: ref
  }, rest, {
    value: maskedValue
  })), inputAccessoryView);
};

exports.MaskedTextInputComponent = MaskedTextInputComponent;
const MaskedTextInput = /*#__PURE__*/(0, _react.forwardRef)(MaskedTextInputComponent);
exports.MaskedTextInput = MaskedTextInput;
