// eas-cli/build/credentials/manager/Actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IosActionType = exports.AndroidActionType = exports.Scope = void 0;
var Scope;
(function (Scope) {
    Scope[Scope["Project"] = 0] = "Project";
    Scope[Scope["Manager"] = 1] = "Manager";
    Scope[Scope["Account"] = 2] = "Account";
})(Scope || (exports.Scope = Scope = {}));
var AndroidActionType;
(function (AndroidActionType) {
    AndroidActionType[AndroidActionType["ManageBuildCredentials"] = 0] = "ManageBuildCredentials";
    AndroidActionType[AndroidActionType["ManageFcm"] = 1] = "ManageFcm";
    AndroidActionType[AndroidActionType["ManageGoogleServiceAccountKeyForSubmissions"] = 2] = "ManageGoogleServiceAccountKeyForSubmissions";
    AndroidActionType[AndroidActionType["ManageGoogleServiceAccount"] = 3] = "ManageGoogleServiceAccount";
    AndroidActionType[AndroidActionType["ManageGoogleServiceAccountKeyForFcmV1"] = 4] = "ManageGoogleServiceAccountKeyForFcmV1";
    AndroidActionType[AndroidActionType["ManageCredentialsJson"] = 5] = "ManageCredentialsJson";
    AndroidActionType[AndroidActionType["GoBackToCaller"] = 6] = "GoBackToCaller";
    AndroidActionType[AndroidActionType["GoBackToHighLevelActions"] = 7] = "GoBackToHighLevelActions";
    AndroidActionType[AndroidActionType["CreateKeystore"] = 8] = "CreateKeystore";
    AndroidActionType[AndroidActionType["SetDefaultKeystore"] = 9] = "SetDefaultKeystore";
    AndroidActionType[AndroidActionType["DownloadKeystore"] = 10] = "DownloadKeystore";
    AndroidActionType[AndroidActionType["RemoveKeystore"] = 11] = "RemoveKeystore";
    AndroidActionType[AndroidActionType["CreateFcm"] = 12] = "CreateFcm";
    AndroidActionType[AndroidActionType["RemoveFcm"] = 13] = "RemoveFcm";
    AndroidActionType[AndroidActionType["CreateGsaKey"] = 14] = "CreateGsaKey";
    AndroidActionType[AndroidActionType["UseExistingGsaKeyForSubmissions"] = 15] = "UseExistingGsaKeyForSubmissions";
    AndroidActionType[AndroidActionType["RemoveGsaKey"] = 16] = "RemoveGsaKey";
    AndroidActionType[AndroidActionType["SetUpGsaKeyForSubmissions"] = 17] = "SetUpGsaKeyForSubmissions";
    AndroidActionType[AndroidActionType["CreateGsaKeyForFcmV1"] = 18] = "CreateGsaKeyForFcmV1";
    AndroidActionType[AndroidActionType["UseExistingGsaKeyForFcmV1"] = 19] = "UseExistingGsaKeyForFcmV1";
    AndroidActionType[AndroidActionType["RemoveGsaKeyForFcmV1"] = 20] = "RemoveGsaKeyForFcmV1";
    AndroidActionType[AndroidActionType["SetUpGsaKeyForFcmV1"] = 21] = "SetUpGsaKeyForFcmV1";
    AndroidActionType[AndroidActionType["UpdateCredentialsJson"] = 22] = "UpdateCredentialsJson";
    AndroidActionType[AndroidActionType["SetUpBuildCredentialsFromCredentialsJson"] = 23] = "SetUpBuildCredentialsFromCredentialsJson";
    AndroidActionType[AndroidActionType["SetUpBuildCredentials"] = 24] = "SetUpBuildCredentials";
})(AndroidActionType || (exports.AndroidActionType = AndroidActionType = {}));
var IosActionType;
(function (IosActionType) {
    IosActionType[IosActionType["ManageCredentialsJson"] = 0] = "ManageCredentialsJson";
    IosActionType[IosActionType["ManageBuildCredentials"] = 1] = "ManageBuildCredentials";
    IosActionType[IosActionType["ManagePushKey"] = 2] = "ManagePushKey";
    IosActionType[IosActionType["ManageAscApiKey"] = 3] = "ManageAscApiKey";
    IosActionType[IosActionType["GoBackToCaller"] = 4] = "GoBackToCaller";
    IosActionType[IosActionType["GoBackToHighLevelActions"] = 5] = "GoBackToHighLevelActions";
    IosActionType[IosActionType["SetUpBuildCredentials"] = 6] = "SetUpBuildCredentials";
    IosActionType[IosActionType["SetUpBuildCredentialsFromCredentialsJson"] = 7] = "SetUpBuildCredentialsFromCredentialsJson";
    IosActionType[IosActionType["UpdateCredentialsJson"] = 8] = "UpdateCredentialsJson";
    IosActionType[IosActionType["UseExistingDistributionCertificate"] = 9] = "UseExistingDistributionCertificate";
    IosActionType[IosActionType["RemoveProvisioningProfile"] = 10] = "RemoveProvisioningProfile";
    IosActionType[IosActionType["CreateDistributionCertificate"] = 11] = "CreateDistributionCertificate";
    IosActionType[IosActionType["RemoveDistributionCertificate"] = 12] = "RemoveDistributionCertificate";
    IosActionType[IosActionType["SetUpPushKey"] = 13] = "SetUpPushKey";
    IosActionType[IosActionType["CreatePushKey"] = 14] = "CreatePushKey";
    IosActionType[IosActionType["UseExistingPushKey"] = 15] = "UseExistingPushKey";
    IosActionType[IosActionType["RemovePushKey"] = 16] = "RemovePushKey";
    IosActionType[IosActionType["SetUpAscApiKeyForSubmissions"] = 17] = "SetUpAscApiKeyForSubmissions";
    IosActionType[IosActionType["UseExistingAscApiKeyForSubmissions"] = 18] = "UseExistingAscApiKeyForSubmissions";
    IosActionType[IosActionType["CreateAscApiKeyForSubmissions"] = 19] = "CreateAscApiKeyForSubmissions";
    IosActionType[IosActionType["RemoveAscApiKey"] = 20] = "RemoveAscApiKey";
})(IosActionType || (exports.IosActionType = IosActionType = {}));
