// eas-cli/build/credentials/manager/AndroidActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gsaActions = exports.gsaKeyActionsForSubmissions = exports.gsaKeyActionsForFcmV1 = exports.fcmActions = exports.buildCredentialsActions = exports.credentialsJsonActions = exports.highLevelActions = void 0;
const Actions_1 = require("./Actions");
exports.highLevelActions = [
    {
        value: Actions_1.AndroidActionType.ManageBuildCredentials,
        title: 'Keystore: Manage everything needed to build your project',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageGoogleServiceAccount,
        title: 'Google Service Account',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageFcm,
        title: 'Push Notifications (Legacy): Manage your FCM (Legacy) API Key',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageCredentialsJson,
        title: 'credentials.json: Upload/Download credentials between EAS servers and your local json ',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToCaller,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.credentialsJsonActions = [
    {
        value: Actions_1.AndroidActionType.UpdateCredentialsJson,
        title: 'Download credentials from EAS to credentials.json',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.SetUpBuildCredentialsFromCredentialsJson,
        title: 'Upload credentials from credentials.json to EAS',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.buildCredentialsActions = [
    {
        value: Actions_1.AndroidActionType.CreateKeystore,
        title: 'Set up a new keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.SetDefaultKeystore,
        title: 'Change default keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.DownloadKeystore,
        title: 'Download existing keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.RemoveKeystore,
        title: 'Delete your keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.fcmActions = [
    {
        value: Actions_1.AndroidActionType.CreateFcm,
        title: 'Upload an FCM API Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.RemoveFcm,
        title: 'Delete your FCM API Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.gsaKeyActionsForFcmV1 = [
    {
        value: Actions_1.AndroidActionType.SetUpGsaKeyForFcmV1,
        title: 'Set up a Google Service Account Key for Push Notifications (FCM V1)',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.UseExistingGsaKeyForFcmV1,
        title: 'Select an existing Google Service Account Key for Push Notifications (FCM V1)',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.gsaKeyActionsForSubmissions = [
    {
        value: Actions_1.AndroidActionType.SetUpGsaKeyForSubmissions,
        title: 'Set up a Google Service Account Key for Play Store Submissions',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.UseExistingGsaKeyForSubmissions,
        title: 'Select an existing Google Service Account Key for Play Store Submissions',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.gsaActions = [
    {
        value: Actions_1.AndroidActionType.ManageGoogleServiceAccountKeyForSubmissions,
        title: 'Manage your Google Service Account Key for Play Store Submissions',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageGoogleServiceAccountKeyForFcmV1,
        title: 'Manage your Google Service Account Key for Push Notifications (FCM V1)',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.CreateGsaKey,
        title: 'Upload a Google Service Account Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.RemoveGsaKey,
        title: 'Delete a Google Service Account Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
