// eas-cli/build/credentials/manager/CheckBuildProfileFlagAgainstEasJson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckBuildProfileFlagAgainstEasJson = void 0;
const tslib_1 = require("tslib");
const eas_json_1 = require("@expo/eas-json");
const log_1 = tslib_1.__importDefault(require("../../log"));
class CheckBuildProfileFlagAgainstEasJson {
    constructor(projectDir, platform, profileNameFromFlag) {
        this.platform = platform;
        this.profileNameFromFlag = profileNameFromFlag;
        this.easJsonAccessor = eas_json_1.EasJsonAccessor.fromProjectPath(projectDir);
    }
    async runAsync() {
        const profileName = await this.getProfileNameFromEasConfigAsync();
        const easConfig = await eas_json_1.EasJsonUtils.getBuildProfileAsync(this.easJsonAccessor, this.platform, profileName);
        log_1.default.succeed(`Using build profile: ${profileName}`);
        return easConfig;
    }
    async getProfileNameFromEasConfigAsync() {
        const buildProfileNames = await eas_json_1.EasJsonUtils.getBuildProfileNamesAsync(this.easJsonAccessor);
        if (buildProfileNames.length === 0) {
            throw new Error('You need at least one iOS build profile declared in eas.json. Go to https://docs.expo.dev/build/eas-json/ for more details');
        }
        else if (buildProfileNames.length === 1) {
            return buildProfileNames[0];
        }
        if (buildProfileNames.includes(this.profileNameFromFlag)) {
            return this.profileNameFromFlag;
        }
        else {
            throw new Error(`Build profile ${this.profileNameFromFlag} does not exist in eas.json. Go to https://docs.expo.dev/build/eas-json/ for more details`);
        }
    }
}
exports.CheckBuildProfileFlagAgainstEasJson = CheckBuildProfileFlagAgainstEasJson;
