// eas-cli/build/credentials/manager/CreateAndroidBuildCredentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateAndroidBuildCredentials = void 0;
const tslib_1 = require("tslib");
const log_1 = tslib_1.__importDefault(require("../../log"));
const prompts_1 = require("../../prompts");
const BuildCredentialsUtils_1 = require("../android/actions/BuildCredentialsUtils");
const CreateKeystore_1 = require("../android/actions/CreateKeystore");
class CreateAndroidBuildCredentials {
    constructor(app) {
        this.app = app;
    }
    async runAsync(ctx) {
        const name = await (0, BuildCredentialsUtils_1.promptForNameAsync)();
        const buildCredentialsList = await ctx.android.getAndroidAppBuildCredentialsListAsync(ctx.graphqlClient, this.app);
        let isDefault = true;
        if (buildCredentialsList.length > 0) {
            isDefault = await (0, prompts_1.confirmAsync)({
                message: 'Do you want to set this as your default build credentials?',
                initial: false,
            });
        }
        const keystore = await new CreateKeystore_1.CreateKeystore(this.app.account).runAsync(ctx);
        const createAndroidCredentials = await ctx.android.createAndroidAppBuildCredentialsAsync(ctx.graphqlClient, this.app, {
            name,
            isDefault,
            androidKeystoreId: keystore.id,
        });
        log_1.default.succeed(`Created Android build credentials ${createAndroidCredentials.name}`);
        return createAndroidCredentials;
    }
}
exports.CreateAndroidBuildCredentials = CreateAndroidBuildCredentials;
