// eas-cli/build/credentials/manager/SetUpBuildCredentialsCommandAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetUpBuildCredentialsCommandAction = void 0;
const eas_build_job_1 = require("@expo/eas-build-job");
const SetUpAndroidBuildCredentials_1 = require("../manager/SetUpAndroidBuildCredentials");
const SetUpIosBuildCredentials_1 = require("../manager/SetUpIosBuildCredentials");
class SetUpBuildCredentialsCommandAction {
    constructor(actor, graphqlClient, vcsClient, analytics, projectInfo, getDynamicPrivateProjectConfigAsync, platform, profileName, projectDir) {
        this.actor = actor;
        this.graphqlClient = graphqlClient;
        this.vcsClient = vcsClient;
        this.analytics = analytics;
        this.projectInfo = projectInfo;
        this.getDynamicPrivateProjectConfigAsync = getDynamicPrivateProjectConfigAsync;
        this.platform = platform;
        this.profileName = profileName;
        this.projectDir = projectDir;
    }
    async runAsync() {
        if (this.platform === eas_build_job_1.Platform.IOS) {
            return await new SetUpIosBuildCredentials_1.SetUpIosBuildCredentials(this, this.projectDir, this.profileName).runAsync();
        }
        return await new SetUpAndroidBuildCredentials_1.SetUpAndroidBuildCredentials(this, this.projectDir, this.profileName).runAsync();
    }
}
exports.SetUpBuildCredentialsCommandAction = SetUpBuildCredentialsCommandAction;
