// @sentry/react-native/dist/js/utils/rnlibraries.js
/* eslint-disable @typescript-eslint/no-var-requires */
import { TurboModuleRegistry } from 'react-native';
import { version as RNV } from 'react-native/Libraries/Core/ReactNativeVersion';
export const ReactNativeLibraries = {
    Devtools: {
        parseErrorStack: (errorStack) => {
            const parseErrorStack = require('react-native/Libraries/Core/Devtools/parseErrorStack');
            return parseErrorStack(errorStack);
        },
        symbolicateStackTrace: (stack, extraData) => {
            const symbolicateStackTrace = require('react-native/Libraries/Core/Devtools/symbolicateStackTrace');
            return symbolicateStackTrace(stack, extraData);
        },
        getDevServer: () => {
            const getDevServer = require('react-native/Libraries/Core/Devtools/getDevServer');
            return getDevServer();
        },
    },
    Promise: require('react-native/Libraries/Promise'),
    Utilities: {
        polyfillGlobal: (name, getValue) => {
            const { polyfillGlobal } = require('react-native/Libraries/Utilities/PolyfillFunctions');
            polyfillGlobal(name, getValue);
        },
    },
    ReactNativeVersion: {
        version: RNV,
    },
    TurboModuleRegistry,
};
