// eas-cli/build/build/ios/prepareJob.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareCredentialsToResign = exports.prepareJobAsync = void 0;
const tslib_1 = require("tslib");
const eas_build_job_1 = require("@expo/eas-build-job");
const nullthrows_1 = tslib_1.__importDefault(require("nullthrows"));
const path_1 = tslib_1.__importDefault(require("path"));
const slash_1 = tslib_1.__importDefault(require("slash"));
const customBuildConfig_1 = require("../../project/customBuildConfig");
const projectUtils_1 = require("../../project/projectUtils");
const cacheDefaults = {
    disabled: false,
    paths: [],
};
async function prepareJobAsync(ctx, jobData) {
    const username = (0, projectUtils_1.getUsername)(ctx.exp, ctx.user);
    const buildProfile = ctx.buildProfile;
    const projectRootDirectory = (0, slash_1.default)(path_1.default.relative(await ctx.vcsClient.getRootPathAsync(), ctx.projectDir)) || '.';
    const buildCredentials = {};
    if (jobData.credentials) {
        const targetNames = Object.keys(jobData.credentials);
        for (const targetName of targetNames) {
            buildCredentials[targetName] = prepareTargetCredentials(jobData.credentials[targetName]);
        }
    }
    const maybeCustomBuildConfigPath = buildProfile.config
        ? (0, customBuildConfig_1.getCustomBuildConfigPathForJob)(buildProfile.config)
        : undefined;
    const job = {
        type: ctx.workflow,
        platform: eas_build_job_1.Platform.IOS,
        projectArchive: jobData.projectArchive,
        projectRootDirectory,
        builderEnvironment: {
            image: buildProfile.image,
            node: buildProfile.node,
            pnpm: buildProfile.pnpm,
            bun: buildProfile.bun,
            yarn: buildProfile.yarn,
            bundler: buildProfile.bundler,
            cocoapods: buildProfile.cocoapods,
            fastlane: buildProfile.fastlane,
            env: buildProfile.env,
        },
        cache: {
            ...cacheDefaults,
            ...buildProfile.cache,
            clear: ctx.clearCache,
        },
        secrets: {
            buildCredentials,
        },
        updates: { channel: buildProfile.channel },
        developmentClient: buildProfile.developmentClient,
        simulator: buildProfile.simulator,
        scheme: jobData.buildScheme,
        buildConfiguration: buildProfile.buildConfiguration,
        applicationArchivePath: buildProfile.applicationArchivePath ?? buildProfile.artifactPath,
        buildArtifactPaths: buildProfile.buildArtifactPaths,
        username,
        ...(ctx.ios.buildNumberOverride && {
            version: {
                buildNumber: ctx.ios.buildNumberOverride,
            },
        }),
        experimental: {
            prebuildCommand: buildProfile.prebuildCommand,
        },
        mode: buildProfile.config || ctx.repack ? eas_build_job_1.BuildMode.CUSTOM : eas_build_job_1.BuildMode.BUILD,
        triggeredBy: eas_build_job_1.BuildTrigger.EAS_CLI,
        ...(maybeCustomBuildConfigPath && {
            customBuildConfig: {
                path: maybeCustomBuildConfigPath,
            },
        }),
        ...(ctx.repack && {
            customBuildConfig: {
                path: '__eas/repack.yml',
            },
        }),
        loggerLevel: ctx.loggerLevel,
    };
    return (0, eas_build_job_1.sanitizeBuildJob)(job);
}
exports.prepareJobAsync = prepareJobAsync;
function prepareCredentialsToResign(credentials) {
    const buildCredentials = [];
    for (const targetName of Object.keys(credentials ?? {})) {
        buildCredentials.push({
            targetName,
            provisioningProfileBase64: (0, nullthrows_1.default)(credentials?.[targetName].provisioningProfile),
            distributionCertificate: {
                dataBase64: (0, nullthrows_1.default)(credentials?.[targetName].distributionCertificate.certificateP12),
                password: (0, nullthrows_1.default)(credentials?.[targetName].distributionCertificate.certificatePassword),
            },
        });
    }
    return { buildCredentials };
}
exports.prepareCredentialsToResign = prepareCredentialsToResign;
function prepareTargetCredentials(targetCredentials) {
    return {
        provisioningProfileBase64: targetCredentials.provisioningProfile,
        distributionCertificate: {
            dataBase64: targetCredentials.distributionCertificate.certificateP12,
            password: targetCredentials.distributionCertificate.certificatePassword,
        },
    };
}
