// @sentry/react-native/dist/js/integrations/sdkinfo.js
import { __awaiter } from "tslib";
import { logger } from '@sentry/utils';
import { isExpoGo, notWeb } from '../utils/environment';
import { SDK_NAME, SDK_PACKAGE_NAME, SDK_VERSION } from '../version';
import { NATIVE } from '../wrapper';
export const defaultSdkInfo = {
    name: SDK_NAME,
    packages: [
        {
            name: SDK_PACKAGE_NAME,
            version: SDK_VERSION,
        },
    ],
    version: SDK_VERSION,
};
/** Default SdkInfo instrumentation */
export class SdkInfo {
    constructor() {
        /**
         * @inheritDoc
         */
        this.name = SdkInfo.id;
        this._nativeSdkPackage = null;
    }
    /**
     * @inheritDoc
     */
    setupOnce(addGlobalEventProcessor) {
        addGlobalEventProcessor((event) => __awaiter(this, void 0, void 0, function* () {
            // this._nativeSdkInfo should be defined a following time so this call won't always be awaited.
            if (this._nativeSdkPackage === null) {
                try {
                    this._nativeSdkPackage = yield NATIVE.fetchNativeSdkInfo();
                }
                catch (e) {
                    // If this fails, go ahead as usual as we would rather have the event be sent with a package missing.
                    if (notWeb() && !isExpoGo()) {
                        logger.warn('[SdkInfo] Native SDK Info retrieval failed...something could be wrong with your Sentry installation:');
                        logger.warn(e);
                    }
                }
            }
            event.platform = event.platform || 'javascript';
            event.sdk = event.sdk || {};
            event.sdk.name = event.sdk.name || defaultSdkInfo.name;
            event.sdk.version = event.sdk.version || defaultSdkInfo.version;
            event.sdk.packages = [
                // default packages are added by baseclient and should not be added here
                ...(event.sdk.packages || []),
                ...((this._nativeSdkPackage && [this._nativeSdkPackage]) || []),
            ];
            return event;
        }));
    }
}
/**
 * @inheritDoc
 */
SdkInfo.id = 'SdkInfo';
