// eas-cli/build/commands/build/configure.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eas_build_job_1 = require("@expo/eas-build-job");
const core_1 = require("@oclif/core");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const syncProjectConfiguration_1 = require("../../build/android/syncProjectConfiguration");
const configure_1 = require("../../build/configure");
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const log_1 = tslib_1.__importStar(require("../../log"));
const platform_1 = require("../../platform");
const projectUtils_1 = require("../../project/projectUtils");
const workflow_1 = require("../../project/workflow");
const prompts_1 = require("../../prompts");
const UpdatesModule_1 = require("../../update/android/UpdatesModule");
const configure_2 = require("../../update/configure");
const UpdatesModule_2 = require("../../update/ios/UpdatesModule");
class BuildConfigure extends EasCommand_1.default {
    async runAsync() {
        const { flags } = await this.parse(_a);
        const { privateProjectConfig: { exp, projectId, projectDir }, vcsClient, } = await this.getContextAsync(_a, {
            nonInteractive: false,
        });
        log_1.default.log('💡 The following process will configure your iOS and/or Android project to be compatible with EAS Build. These changes only apply to your local project files and you can safely revert them at any time.');
        // BuildConfigure.ContextOptions.Vcs.client.getValueAsync()
        await vcsClient.ensureRepoExistsAsync();
        const expoUpdatesIsInstalled = (0, projectUtils_1.isExpoUpdatesInstalled)(projectDir);
        const platform = flags.platform ?? (await promptForPlatformAsync());
        // clean up old Android configuration
        if ([platform_1.RequestedPlatform.Android, platform_1.RequestedPlatform.All].includes(platform)) {
            await (0, syncProjectConfiguration_1.cleanUpOldEasBuildGradleScriptAsync)(projectDir);
        }
        // ensure eas.json exists
        log_1.default.newLine();
        const didCreateEasJson = await (0, configure_1.ensureProjectConfiguredAsync)({
            projectDir,
            nonInteractive: false,
            vcsClient,
        });
        if (didCreateEasJson && (0, projectUtils_1.isUsingEASUpdate)(exp, projectId)) {
            await (0, configure_2.ensureEASUpdateIsConfiguredInEasJsonAsync)(projectDir);
        }
        // configure expo-updates
        if (expoUpdatesIsInstalled) {
            if ([platform_1.RequestedPlatform.Android, platform_1.RequestedPlatform.All].includes(platform)) {
                const workflow = await (0, workflow_1.resolveWorkflowAsync)(projectDir, eas_build_job_1.Platform.ANDROID, vcsClient);
                if (workflow === eas_build_job_1.Workflow.GENERIC) {
                    await (0, UpdatesModule_1.syncUpdatesConfigurationAsync)({ projectDir, exp, workflow, env: undefined });
                }
            }
            if ([platform_1.RequestedPlatform.Ios, platform_1.RequestedPlatform.All].includes(platform)) {
                const workflow = await (0, workflow_1.resolveWorkflowAsync)(projectDir, eas_build_job_1.Platform.IOS, vcsClient);
                if (workflow === eas_build_job_1.Workflow.GENERIC) {
                    await (0, UpdatesModule_2.syncUpdatesConfigurationAsync)({
                        vcsClient,
                        projectDir,
                        exp,
                        workflow,
                        env: undefined,
                    });
                }
            }
        }
        log_1.default.addNewLineIfNone();
        log_1.default.log(`🎉 Your project is ready to build.

- Run ${chalk_1.default.bold('eas build')} when you are ready to create your first build.
- Once the build is completed, run ${chalk_1.default.bold('eas submit')} to upload the app to app stores.
- ${(0, log_1.learnMore)('https://docs.expo.dev/build/introduction', {
            learnMoreMessage: 'Learn more about other capabilities of EAS Build',
        })}`);
    }
}
_a = BuildConfigure;
BuildConfigure.description = 'configure the project to support EAS Build';
BuildConfigure.flags = {
    platform: core_1.Flags.enum({
        description: 'Platform to configure',
        char: 'p',
        options: ['android', 'ios', 'all'],
    }),
};
BuildConfigure.contextDefinition = {
    ..._a.ContextOptions.ProjectConfig,
    ..._a.ContextOptions.LoggedIn,
    ..._a.ContextOptions.Vcs,
};
exports.default = BuildConfigure;
async function promptForPlatformAsync() {
    log_1.default.addNewLineIfNone();
    const { platform } = await (0, prompts_1.promptAsync)({
        type: 'select',
        message: 'Which platforms would you like to configure for EAS Build?',
        name: 'platform',
        choices: [
            {
                title: 'All',
                value: platform_1.RequestedPlatform.All,
            },
            {
                title: 'iOS',
                value: platform_1.RequestedPlatform.Ios,
            },
            {
                title: 'Android',
                value: platform_1.RequestedPlatform.Android,
            },
        ],
    });
    return platform;
}
