// eas-cli/build/commands/build/delete.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectBuildToDeleteAsync = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const graphql_tag_1 = tslib_1.__importDefault(require("graphql-tag"));
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const builds_1 = require("../../commandUtils/builds");
const flags_1 = require("../../commandUtils/flags");
const client_1 = require("../../graphql/client");
const log_1 = tslib_1.__importDefault(require("../../log"));
const ora_1 = require("../../ora");
const platform_1 = require("../../platform");
const projectUtils_1 = require("../../project/projectUtils");
const prompts_1 = require("../../prompts");
async function deleteBuildAsync(graphqlClient, buildId) {
    const data = await (0, client_1.withErrorHandlingAsync)(graphqlClient
        .mutation((0, graphql_tag_1.default) `
          mutation DeleteBuildMutation($buildId: ID!) {
            build(buildId: $buildId) {
              deleteBuild(buildId: $buildId) {
                id
              }
            }
          }
        `, { buildId })
        .toPromise());
    return data.build.deleteBuild;
}
async function selectBuildToDeleteAsync(graphqlClient, projectId, projectDisplayName, filters) {
    const spinner = (0, ora_1.ora)().start('Fetching builds…');
    let builds;
    try {
        builds = await (0, builds_1.fetchBuildsAsync)({ graphqlClient, projectId, filters });
        spinner.stop();
    }
    catch (error) {
        spinner.fail(`Something went wrong and we couldn't fetch the builds for the project ${projectDisplayName}.`);
        throw error;
    }
    if (builds.length === 0) {
        log_1.default.warn(`We couldn't find any builds for the project ${projectDisplayName}.`);
        return null;
    }
    else {
        const buildId = await (0, prompts_1.selectAsync)('Which build do you want to delete?', builds.map(build => ({
            title: (0, builds_1.formatBuild)(build),
            value: build.id,
        })));
        return (await (0, prompts_1.confirmAsync)({
            message: 'Are you sure you want to delete it?',
        }))
            ? buildId
            : null;
    }
}
exports.selectBuildToDeleteAsync = selectBuildToDeleteAsync;
class BuildDelete extends EasCommand_1.default {
    async runAsync() {
        const { args: { BUILD_ID: buildIdFromArg }, flags: { 'non-interactive': nonInteractive, platform, profile }, } = await this.parse(_a);
        if (buildIdFromArg && (platform || profile)) {
            throw new Error('Build ID cannot be used together with platform and profile flags. They are used to filter the list of builds when not providing the build ID');
        }
        const { privateProjectConfig: { projectId }, loggedIn: { graphqlClient }, } = await this.getContextAsync(_a, { nonInteractive });
        const displayName = await (0, projectUtils_1.getDisplayNameForProjectIdAsync)(graphqlClient, projectId);
        if (buildIdFromArg) {
            await (0, builds_1.ensureBuildExistsAsync)(graphqlClient, buildIdFromArg);
        }
        let buildId = buildIdFromArg;
        if (!buildId) {
            if (nonInteractive) {
                throw new Error('Build ID must be provided in non-interactive mode');
            }
            buildId = await selectBuildToDeleteAsync(graphqlClient, projectId, displayName, {
                platform,
                profile,
            });
            if (!buildId) {
                return;
            }
        }
        const spinner = (0, ora_1.ora)().start('Deleting the build...');
        try {
            await deleteBuildAsync(graphqlClient, buildId);
            spinner.succeed('Build deleted');
        }
        catch (error) {
            spinner.fail(`Something went wrong and we couldn't delete your build ${buildId}`);
            throw error;
        }
    }
}
_a = BuildDelete;
BuildDelete.description = 'delete a build';
BuildDelete.args = [{ name: 'BUILD_ID' }];
BuildDelete.flags = {
    ...flags_1.EASNonInteractiveFlag,
    platform: core_1.Flags.enum({
        char: 'p',
        description: 'Filter builds by the platform if build ID is not provided',
        options: Object.values(platform_1.RequestedPlatform),
    }),
    profile: core_1.Flags.string({
        char: 'e',
        description: 'Filter builds by build profile if build ID is not provided',
        helpValue: 'PROFILE_NAME',
    }),
};
BuildDelete.contextDefinition = {
    ..._a.ContextOptions.ProjectConfig,
    ..._a.ContextOptions.LoggedIn,
    ..._a.ContextOptions.Vcs,
};
exports.default = BuildDelete;
