// eas-cli/build/graphql/queries/BuildQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuildQuery = void 0;
const tslib_1 = require("tslib");
const graphql_1 = require("graphql");
const graphql_tag_1 = tslib_1.__importDefault(require("graphql-tag"));
const client_1 = require("../client");
const Build_1 = require("../types/Build");
exports.BuildQuery = {
    async byIdAsync(graphqlClient, buildId, { useCache = true } = {}) {
        const data = await (0, client_1.withErrorHandlingAsync)(graphqlClient
            .query((0, graphql_tag_1.default) `
            query BuildsByIdQuery($buildId: ID!) {
              builds {
                byId(buildId: $buildId) {
                  id
                  ...BuildFragment
                }
              }
            }
            ${(0, graphql_1.print)(Build_1.BuildFragmentNode)}
          `, { buildId }, {
            requestPolicy: useCache ? 'cache-first' : 'network-only',
            additionalTypenames: ['Build'],
        })
            .toPromise());
        return data.builds.byId;
    },
    async withSubmissionsByIdAsync(graphqlClient, buildId, { useCache = true } = {}) {
        const data = await (0, client_1.withErrorHandlingAsync)(graphqlClient
            .query((0, graphql_tag_1.default) `
            query BuildsWithSubmissionsByIdQuery($buildId: ID!) {
              builds {
                byId(buildId: $buildId) {
                  id
                  ...BuildWithSubmissionsFragment
                }
              }
            }
            ${(0, graphql_1.print)(Build_1.BuildFragmentWithSubmissionsNode)}
          `, { buildId }, {
            requestPolicy: useCache ? 'cache-first' : 'network-only',
            additionalTypenames: ['Build'],
        })
            .toPromise());
        return data.builds.byId;
    },
    async viewBuildsOnAppAsync(graphqlClient, { appId, limit, offset, filter }) {
        const data = await (0, client_1.withErrorHandlingAsync)(graphqlClient
            .query(
        // TODO: Change $appId: String! to ID! when fixed server-side schema
        (0, graphql_tag_1.default) `
            query ViewBuildsOnApp(
              $appId: String!
              $offset: Int!
              $limit: Int!
              $filter: BuildFilter
            ) {
              app {
                byId(appId: $appId) {
                  id
                  builds(offset: $offset, limit: $limit, filter: $filter) {
                    id
                    ...BuildFragment
                  }
                }
              }
            }
            ${(0, graphql_1.print)(Build_1.BuildFragmentNode)}
          `, { appId, offset, limit, filter }, {
            additionalTypenames: ['Build'],
        })
            .toPromise());
        return data.app?.byId.builds ?? [];
    },
};
