// eas-cli/build/build/android/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformJob = void 0;
const eas_build_job_1 = require("@expo/eas-build-job");
const generated_1 = require("../../graphql/generated");
const graphql_1 = require("../graphql");
function transformJob(job) {
    return {
        type: (0, graphql_1.transformWorkflow)(job.type),
        triggeredBy: (0, graphql_1.transformBuildTrigger)(job.triggeredBy),
        projectArchive: (0, graphql_1.transformProjectArchive)(job.projectArchive),
        projectRootDirectory: job.projectRootDirectory,
        updates: job.updates,
        secrets: job.secrets,
        builderEnvironment: job.builderEnvironment,
        cache: job.cache,
        version: job.version?.versionCode ? { versionCode: job.version.versionCode } : undefined,
        gradleCommand: job.gradleCommand,
        applicationArchivePath: job.applicationArchivePath,
        buildArtifactPaths: job.buildArtifactPaths,
        username: job.username,
        buildType: job.buildType && transformBuildType(job.buildType),
        developmentClient: job.developmentClient,
        experimental: job.experimental,
        mode: (0, graphql_1.transformBuildMode)(job.mode),
        customBuildConfig: job.customBuildConfig,
        loggerLevel: job.loggerLevel
            ? graphql_1.loggerLevelToGraphQLWorkerLoggerLevel[job.loggerLevel]
            : undefined,
    };
}
exports.transformJob = transformJob;
function transformBuildType(buildType) {
    if (buildType === eas_build_job_1.Android.BuildType.APK) {
        return generated_1.AndroidBuildType.Apk;
    }
    else {
        return generated_1.AndroidBuildType.AppBundle;
    }
}
