// eas-cli/build/credentials/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AndroidPackageNotDefinedError = exports.UnsupportedCredentialsChoiceError = exports.MissingCredentialsError = exports.ForbidCredentialModificationError = exports.InsufficientAuthenticationNonInteractiveError = exports.MissingCredentialsNonInteractiveError = void 0;
const log_1 = require("../log");
class MissingCredentialsNonInteractiveError extends Error {
    constructor(message) {
        super(message ?? 'Credentials are not set up. Run this command again in interactive mode.');
    }
}
exports.MissingCredentialsNonInteractiveError = MissingCredentialsNonInteractiveError;
class InsufficientAuthenticationNonInteractiveError extends Error {
    constructor(message) {
        super(message ??
            `Authentication with an ASC API key is required in non-interactive mode. ${(0, log_1.learnMore)('https://docs.expo.dev/build/building-on-ci/#optional-provide-an-asc-api-token-for-your-apple-team')}`);
    }
}
exports.InsufficientAuthenticationNonInteractiveError = InsufficientAuthenticationNonInteractiveError;
class ForbidCredentialModificationError extends Error {
    constructor(message) {
        super(message ??
            'Credentials cannot be modified. Run this command again without the --freeze-credentials flag.');
    }
}
exports.ForbidCredentialModificationError = ForbidCredentialModificationError;
class MissingCredentialsError extends Error {
    constructor(message) {
        super(message ?? 'Credentials are not set up.');
    }
}
exports.MissingCredentialsError = MissingCredentialsError;
class UnsupportedCredentialsChoiceError extends Error {
    constructor(message, choice) {
        super(message);
        this.choice = choice;
    }
}
exports.UnsupportedCredentialsChoiceError = UnsupportedCredentialsChoiceError;
class AndroidPackageNotDefinedError extends Error {
    constructor(message) {
        super(message ?? 'android.package needs to be defined in your app.json/app.config.js file');
    }
}
exports.AndroidPackageNotDefinedError = AndroidPackageNotDefinedError;
