// @expo/steps/dist_esm/cli/cli.js
var _a;
import path from 'path';
import { createLogger } from '@expo/logger';
import { BuildConfigParser } from '../BuildConfigParser.js';
import { BuildStepGlobalContext } from '../BuildStepContext.js';
import { BuildWorkflowError } from '../errors.js';
const logger = createLogger({
    name: 'steps-cli',
    level: 'info',
});
export class CliContextProvider {
    constructor(logger, runtimePlatform, projectSourceDirectory, projectTargetDirectory, defaultWorkingDirectory, buildLogsDirectory) {
        this.logger = logger;
        this.runtimePlatform = runtimePlatform;
        this.projectSourceDirectory = projectSourceDirectory;
        this.projectTargetDirectory = projectTargetDirectory;
        this.defaultWorkingDirectory = defaultWorkingDirectory;
        this.buildLogsDirectory = buildLogsDirectory;
        this._env = {};
    }
    get env() {
        return this._env;
    }
    staticContext() {
        return {
            job: {},
            metadata: {},
            env: this.env,
        };
    }
    updateEnv(env) {
        this._env = env;
    }
}
async function runAsync(configPath, relativeProjectDirectory, runtimePlatform) {
    const ctx = new BuildStepGlobalContext(new CliContextProvider(logger, runtimePlatform, relativeProjectDirectory, relativeProjectDirectory, relativeProjectDirectory, relativeProjectDirectory), false);
    const parser = new BuildConfigParser(ctx, {
        configPath,
    });
    const workflow = await parser.parseAsync();
    await workflow.executeAsync();
}
const relativeConfigPath = process.argv[2];
const relativeProjectDirectoryPath = process.argv[3];
const platform = ((_a = process.argv[4]) !== null && _a !== void 0 ? _a : process.platform);
if (!relativeConfigPath || !relativeProjectDirectoryPath) {
    console.error('Usage: yarn cli config.yml path/to/project/directory [darwin|linux]');
    process.exit(1);
}
const configPath = path.resolve(process.cwd(), relativeConfigPath);
const workingDirectory = path.resolve(process.cwd(), relativeProjectDirectoryPath);
runAsync(configPath, workingDirectory, platform).catch((err) => {
    logger.error({ err }, 'Build failed');
    if (err instanceof BuildWorkflowError) {
        logger.error('Failed to parse the custom build config file.');
        for (const detailedErr of err.errors) {
            logger.error({ err: detailedErr });
        }
    }
});
