// @expo/eas-build-job/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.errors = exports.sanitizeMetadata = exports.Platform = exports.Workflow = exports.EnvironmentSecretType = exports.BuildTrigger = exports.BuildMode = exports.ArchiveSourceType = exports.Ios = exports.Android = void 0;
exports.Android = __importStar(require("./android"));
exports.Ios = __importStar(require("./ios"));
var common_1 = require("./common");
Object.defineProperty(exports, "ArchiveSourceType", { enumerable: true, get: function () { return common_1.ArchiveSourceType; } });
Object.defineProperty(exports, "BuildMode", { enumerable: true, get: function () { return common_1.BuildMode; } });
Object.defineProperty(exports, "BuildTrigger", { enumerable: true, get: function () { return common_1.BuildTrigger; } });
Object.defineProperty(exports, "EnvironmentSecretType", { enumerable: true, get: function () { return common_1.EnvironmentSecretType; } });
Object.defineProperty(exports, "Workflow", { enumerable: true, get: function () { return common_1.Workflow; } });
Object.defineProperty(exports, "Platform", { enumerable: true, get: function () { return common_1.Platform; } });
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "sanitizeMetadata", { enumerable: true, get: function () { return metadata_1.sanitizeMetadata; } });
__exportStar(require("./job"), exports);
__exportStar(require("./logs"), exports);
exports.errors = __importStar(require("./errors"));
__exportStar(require("./artifacts"), exports);
__exportStar(require("./context"), exports);
__exportStar(require("./generic"), exports);
