// @expo/eas-build-job/dist/logs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogMarker = exports.BuildPhaseResult = exports.XCODE_LOGS_BUILD_PHASE_WEBSITE_ID = exports.submissionPhaseWebsiteId = exports.buildPhaseWebsiteId = exports.submissionPhaseDisplayName = exports.buildPhaseDisplayName = exports.SubmissionPhase = exports.BuildPhase = void 0;
var BuildPhase;
(function (BuildPhase) {
    BuildPhase["UNKNOWN"] = "UNKNOWN";
    BuildPhase["QUEUE"] = "QUEUE";
    BuildPhase["SPIN_UP_BUILDER"] = "SPIN_UP_BUILDER";
    BuildPhase["BUILDER_INFO"] = "BUILDER_INFO";
    BuildPhase["READ_APP_CONFIG"] = "READ_APP_CONFIG";
    BuildPhase["READ_PACKAGE_JSON"] = "READ_PACKAGE_JSON";
    BuildPhase["RUN_EXPO_DOCTOR"] = "RUN_EXPO_DOCTOR";
    BuildPhase["SET_UP_BUILD_ENVIRONMENT"] = "SET_UP_BUILD_ENVIRONMENT";
    BuildPhase["START_BUILD"] = "START_BUILD";
    BuildPhase["INSTALL_CUSTOM_TOOLS"] = "INSTALL_CUSTOM_TOOLS";
    BuildPhase["PREPARE_PROJECT"] = "PREPARE_PROJECT";
    BuildPhase["RESTORE_CACHE"] = "RESTORE_CACHE";
    BuildPhase["INSTALL_DEPENDENCIES"] = "INSTALL_DEPENDENCIES";
    BuildPhase["EAS_BUILD_INTERNAL"] = "EAS_BUILD_INTERNAL";
    BuildPhase["PREBUILD"] = "PREBUILD";
    BuildPhase["PREPARE_CREDENTIALS"] = "PREPARE_CREDENTIALS";
    BuildPhase["CALCULATE_EXPO_UPDATES_RUNTIME_VERSION"] = "CALCULATE_EXPO_UPDATES_RUNTIME_VERSION";
    BuildPhase["CONFIGURE_EXPO_UPDATES"] = "CONFIGURE_EXPO_UPDATES";
    BuildPhase["SAVE_CACHE"] = "SAVE_CACHE";
    /**
     * @deprecated
     */
    BuildPhase["UPLOAD_ARTIFACTS"] = "UPLOAD_ARTIFACTS";
    BuildPhase["UPLOAD_APPLICATION_ARCHIVE"] = "UPLOAD_APPLICATION_ARCHIVE";
    BuildPhase["UPLOAD_BUILD_ARTIFACTS"] = "UPLOAD_BUILD_ARTIFACTS";
    BuildPhase["PREPARE_ARTIFACTS"] = "PREPARE_ARTIFACTS";
    BuildPhase["CLEAN_UP_CREDENTIALS"] = "CLEAN_UP_CREDENTIALS";
    BuildPhase["COMPLETE_BUILD"] = "COMPLETE_BUILD";
    BuildPhase["FAIL_BUILD"] = "FAIL_BUILD";
    // ANDROID
    BuildPhase["FIX_GRADLEW"] = "FIX_GRADLEW";
    BuildPhase["RUN_GRADLEW"] = "RUN_GRADLEW";
    // IOS
    BuildPhase["INSTALL_PODS"] = "INSTALL_PODS";
    BuildPhase["CONFIGURE_XCODE_PROJECT"] = "CONFIGURE_XCODE_PROJECT";
    BuildPhase["RUN_FASTLANE"] = "RUN_FASTLANE";
    // HOOKS
    BuildPhase["PRE_INSTALL_HOOK"] = "PRE_INSTALL_HOOK";
    BuildPhase["POST_INSTALL_HOOK"] = "POST_INSTALL_HOOK";
    BuildPhase["PRE_UPLOAD_ARTIFACTS_HOOK"] = "PRE_UPLOAD_ARTIFACTS_HOOK";
    BuildPhase["ON_BUILD_SUCCESS_HOOK"] = "ON_BUILD_SUCCESS_HOOK";
    BuildPhase["ON_BUILD_ERROR_HOOK"] = "ON_BUILD_ERROR_HOOK";
    BuildPhase["ON_BUILD_COMPLETE_HOOK"] = "ON_BUILD_COMPLETE_HOOK";
    BuildPhase["ON_BUILD_CANCEL_HOOK"] = "ON_BUILD_CANCEL_HOOK";
    // RESIGN
    BuildPhase["DOWNLOAD_APPLICATION_ARCHIVE"] = "DOWNLOAD_APPLICATION_ARCHIVE";
    // CUSTOM BUILDS
    BuildPhase["PARSE_CUSTOM_WORKFLOW_CONFIG"] = "PARSE_CUSTOM_WORKFLOW_CONFIG";
    BuildPhase["CUSTOM"] = "CUSTOM";
})(BuildPhase || (exports.BuildPhase = BuildPhase = {}));
var SubmissionPhase;
(function (SubmissionPhase) {
    SubmissionPhase["SPIN_UP_SUBMISSION_WORKER"] = "SPIN_UP_SUBMISSION_WORKER";
    SubmissionPhase["SUBMIT_TO_PLAY_STORE"] = "SUBMIT_TO_PLAY_STORE";
    SubmissionPhase["SUBMIT_TO_APP_STORE"] = "SUBMIT_TO_APP_STORE";
    SubmissionPhase["FAIL_SUBMISSION"] = "FAIL_SUBMISSION";
})(SubmissionPhase || (exports.SubmissionPhase = SubmissionPhase = {}));
exports.buildPhaseDisplayName = {
    [BuildPhase.UNKNOWN]: 'Unknown build phase',
    [BuildPhase.QUEUE]: 'Waiting to start',
    [BuildPhase.SPIN_UP_BUILDER]: 'Spin up build environment',
    [BuildPhase.SET_UP_BUILD_ENVIRONMENT]: 'Set up build environment',
    [BuildPhase.BUILDER_INFO]: 'Builder environment info',
    [BuildPhase.START_BUILD]: 'Start build',
    [BuildPhase.INSTALL_CUSTOM_TOOLS]: 'Install custom tools',
    [BuildPhase.PREPARE_PROJECT]: 'Prepare project',
    [BuildPhase.RESTORE_CACHE]: 'Restore cache',
    [BuildPhase.INSTALL_DEPENDENCIES]: 'Install dependencies',
    [BuildPhase.EAS_BUILD_INTERNAL]: 'Resolve build configuration',
    [BuildPhase.PREBUILD]: 'Prebuild',
    [BuildPhase.PREPARE_CREDENTIALS]: 'Prepare credentials',
    [BuildPhase.CALCULATE_EXPO_UPDATES_RUNTIME_VERSION]: 'Calculate expo-updates runtime version',
    [BuildPhase.CONFIGURE_EXPO_UPDATES]: 'Configure expo-updates',
    [BuildPhase.SAVE_CACHE]: 'Save cache',
    [BuildPhase.UPLOAD_ARTIFACTS]: 'Upload artifacts',
    [BuildPhase.UPLOAD_APPLICATION_ARCHIVE]: 'Upload application archive',
    [BuildPhase.UPLOAD_BUILD_ARTIFACTS]: 'Upload build artifacts',
    [BuildPhase.PREPARE_ARTIFACTS]: 'Prepare artifacts',
    [BuildPhase.CLEAN_UP_CREDENTIALS]: 'Clean up credentials',
    [BuildPhase.COMPLETE_BUILD]: 'Complete build',
    [BuildPhase.FAIL_BUILD]: 'Fail build',
    [BuildPhase.READ_APP_CONFIG]: 'Read app config',
    [BuildPhase.READ_PACKAGE_JSON]: 'Read package.json',
    [BuildPhase.RUN_EXPO_DOCTOR]: 'Run expo doctor',
    [BuildPhase.DOWNLOAD_APPLICATION_ARCHIVE]: 'Download application archive',
    // ANDROID
    [BuildPhase.FIX_GRADLEW]: 'Fix gradlew',
    [BuildPhase.RUN_GRADLEW]: 'Run gradlew',
    // IOS
    [BuildPhase.INSTALL_PODS]: 'Install pods',
    [BuildPhase.CONFIGURE_XCODE_PROJECT]: 'Configure Xcode project',
    [BuildPhase.RUN_FASTLANE]: 'Run fastlane',
    // HOOKS
    [BuildPhase.PRE_INSTALL_HOOK]: 'Pre-install hook',
    [BuildPhase.POST_INSTALL_HOOK]: 'Post-install hook',
    [BuildPhase.PRE_UPLOAD_ARTIFACTS_HOOK]: 'Pre-upload-artifacts hook',
    [BuildPhase.ON_BUILD_SUCCESS_HOOK]: 'Build success hook',
    [BuildPhase.ON_BUILD_ERROR_HOOK]: 'Build error hook',
    [BuildPhase.ON_BUILD_COMPLETE_HOOK]: 'Build complete hook',
    [BuildPhase.ON_BUILD_CANCEL_HOOK]: 'Build cancel hook',
    // CUSTOM
    [BuildPhase.CUSTOM]: 'Unknown build phase',
    [BuildPhase.PARSE_CUSTOM_WORKFLOW_CONFIG]: 'Parse custom build config',
};
exports.submissionPhaseDisplayName = {
    [SubmissionPhase.SPIN_UP_SUBMISSION_WORKER]: 'Spin up submission worker',
    [SubmissionPhase.SUBMIT_TO_PLAY_STORE]: 'Submit to Play Store',
    [SubmissionPhase.SUBMIT_TO_APP_STORE]: 'Submit to App Store',
    [SubmissionPhase.FAIL_SUBMISSION]: 'Fail submission',
};
exports.buildPhaseWebsiteId = {
    [BuildPhase.UNKNOWN]: 'unknown',
    [BuildPhase.QUEUE]: 'waiting-to-start',
    [BuildPhase.SPIN_UP_BUILDER]: 'spin-up-build-environment',
    [BuildPhase.SET_UP_BUILD_ENVIRONMENT]: 'set-up-build-environment',
    [BuildPhase.BUILDER_INFO]: 'builder-environment-info',
    [BuildPhase.START_BUILD]: 'start-build',
    [BuildPhase.INSTALL_CUSTOM_TOOLS]: 'install-custom-tools',
    [BuildPhase.PREPARE_PROJECT]: 'prepare-project',
    [BuildPhase.RESTORE_CACHE]: 'restore-cache',
    [BuildPhase.INSTALL_DEPENDENCIES]: 'install-dependencies',
    [BuildPhase.EAS_BUILD_INTERNAL]: 'resolve-build-configuration',
    [BuildPhase.PREBUILD]: 'prebuild',
    [BuildPhase.PREPARE_CREDENTIALS]: 'prepare-credentials',
    [BuildPhase.CALCULATE_EXPO_UPDATES_RUNTIME_VERSION]: 'calculate-expo-updates-runtime-version',
    [BuildPhase.CONFIGURE_EXPO_UPDATES]: 'configure-expo-updates',
    [BuildPhase.SAVE_CACHE]: 'save-cache',
    [BuildPhase.UPLOAD_ARTIFACTS]: 'upload-artifacts',
    [BuildPhase.UPLOAD_APPLICATION_ARCHIVE]: 'upload-application-archive',
    [BuildPhase.UPLOAD_BUILD_ARTIFACTS]: 'upload-build-artifacts',
    [BuildPhase.PREPARE_ARTIFACTS]: 'prepare-artifacts',
    [BuildPhase.CLEAN_UP_CREDENTIALS]: 'clean-up-credentials',
    [BuildPhase.COMPLETE_BUILD]: 'complete-build',
    [BuildPhase.FAIL_BUILD]: 'fail-build',
    [BuildPhase.READ_APP_CONFIG]: 'read-app-config',
    [BuildPhase.READ_PACKAGE_JSON]: 'read-package-json',
    [BuildPhase.RUN_EXPO_DOCTOR]: 'run-expo-doctor',
    [BuildPhase.DOWNLOAD_APPLICATION_ARCHIVE]: 'download-application-archive',
    // ANDROID
    [BuildPhase.FIX_GRADLEW]: 'fix-gradlew',
    [BuildPhase.RUN_GRADLEW]: 'run-gradlew',
    // IOS
    [BuildPhase.INSTALL_PODS]: 'install-pods',
    [BuildPhase.CONFIGURE_XCODE_PROJECT]: 'configure-xcode-project',
    [BuildPhase.RUN_FASTLANE]: 'run-fastlane',
    // HOOKS
    [BuildPhase.PRE_INSTALL_HOOK]: 'pre-install-hook',
    [BuildPhase.POST_INSTALL_HOOK]: 'post-install-hook',
    [BuildPhase.PRE_UPLOAD_ARTIFACTS_HOOK]: 'pre-upload-artifacts-hook',
    [BuildPhase.ON_BUILD_SUCCESS_HOOK]: 'build-success-hook',
    [BuildPhase.ON_BUILD_ERROR_HOOK]: 'build-error-hook',
    [BuildPhase.ON_BUILD_COMPLETE_HOOK]: 'build-complete-hook',
    [BuildPhase.ON_BUILD_CANCEL_HOOK]: 'build-cancel-hook',
    // CUSTOM
    [BuildPhase.CUSTOM]: 'custom',
    [BuildPhase.PARSE_CUSTOM_WORKFLOW_CONFIG]: 'parse-custom-workflow-config',
};
exports.submissionPhaseWebsiteId = {
    [SubmissionPhase.SPIN_UP_SUBMISSION_WORKER]: 'spin-up-submission-worker',
    [SubmissionPhase.SUBMIT_TO_PLAY_STORE]: 'submit-to-play-store',
    [SubmissionPhase.SUBMIT_TO_APP_STORE]: 'submit-to-app-store',
    [SubmissionPhase.FAIL_SUBMISSION]: 'fail-submission',
};
exports.XCODE_LOGS_BUILD_PHASE_WEBSITE_ID = 'xcode-logs';
var BuildPhaseResult;
(function (BuildPhaseResult) {
    BuildPhaseResult["SUCCESS"] = "success";
    BuildPhaseResult["FAIL"] = "failed";
    BuildPhaseResult["WARNING"] = "warning";
    BuildPhaseResult["SKIPPED"] = "skipped";
    BuildPhaseResult["UNKNOWN"] = "unknown";
})(BuildPhaseResult || (exports.BuildPhaseResult = BuildPhaseResult = {}));
var LogMarker;
(function (LogMarker) {
    LogMarker["START_PHASE"] = "START_PHASE";
    LogMarker["END_PHASE"] = "END_PHASE";
})(LogMarker || (exports.LogMarker = LogMarker = {}));
