// @expo/logger/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipeMode = exports.pipeSpawnOutput = exports.pipe = exports.createLogger = exports.LoggerLevel = void 0;
const bunyan_1 = __importDefault(require("bunyan"));
const level_1 = __importDefault(require("./level"));
exports.LoggerLevel = level_1.default;
const pipe_1 = require("./pipe");
Object.defineProperty(exports, "pipe", { enumerable: true, get: function () { return pipe_1.pipe; } });
Object.defineProperty(exports, "pipeSpawnOutput", { enumerable: true, get: function () { return pipe_1.pipeSpawnOutput; } });
Object.defineProperty(exports, "PipeMode", { enumerable: true, get: function () { return pipe_1.PipeMode; } });
const DEFAULT_LOGGER_NAME = 'expo-logger';
function createLogger(options) {
    return bunyan_1.default.createLogger({
        serializers: bunyan_1.default.stdSerializers,
        ...options,
    });
}
exports.createLogger = createLogger;
const defaultLogger = createLogger({
    name: DEFAULT_LOGGER_NAME,
    level: level_1.default.INFO,
});
exports.default = defaultLogger;
